File: scripts/fetch-mcp-data.js
const fs = require('fs');
const https = require('https');

const fetchMCPData = async (pageNumber, pageSize = 30) => {
  return new Promise((resolve, reject) => {
    const postData = JSON.stringify({
      PageSize: pageSize,
      PageNumber: pageNumber,
      Query: "",
      Criterion: []
    });

    const options = {
      hostname: 'modelscope.cn',
      port: 443,
      path: '/api/v1/dolphin/mcpServers',
      method: 'PUT',
      headers: {
        'Accept': 'application/json, text/plain, */*',
        'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData),
        'Origin': 'https://modelscope.cn',
        'Referer': `https://modelscope.cn/mcp?page=${pageNumber}`,
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/141.0.0.0 Safari/537.36',
      }
    };

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        try {
          const jsonData = JSON.parse(data);
          resolve(jsonData);
        } catch (error) {
          reject(error);
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.write(postData);
    req.end();
  });
};

const fetchAllPages = async () => {
  console.log('开始获取MCP服务数据...');

  let allData = [];
  let currentPage = 1;
  let hasMore = true;

  while (hasMore) {
    try {
      console.log(`正在获取第 ${currentPage} 页...`);
      const result = await fetchMCPData(currentPage);

      if (result.Data && result.Data.Servers && result.Data.Servers.length > 0) {
        allData = allData.concat(result.Data.Servers);
        console.log(`第 ${currentPage} 页获取成功，获得 ${result.Data.Servers.length} 条数据`);

        // 检查是否还有更多数据
        const totalCount = result.Data.TotalCount || 0;
        const currentTotal = currentPage * 30;

        if (currentTotal >= totalCount) {
          hasMore = false;
          console.log(`已获取全部数据，总计 ${allData.length} 条`);
        } else {
          currentPage++;
          // 添加延迟避免请求过快
          await new Promise(resolve => setTimeout(resolve, 1000));
        }
      } else {
        hasMore = false;
        console.log('没有更多数据');
      }
    } catch (error) {
      console.error(`获取第 ${currentPage} 页时出错:`, error.message);
      hasMore = false;
    }
  }

  // 保存到文件
  const outputPath = './src/data/mcp-services.json';
  const dirPath = './src/data';

  // 确保目录存在
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }

  fs.writeFileSync(outputPath, JSON.stringify(allData, null, 2), 'utf-8');
  console.log(`数据已保存到 ${outputPath}`);
  console.log(`总计 ${allData.length} 条MCP服务数据`);

  // 生成统计信息
  const stats = {
    total: allData.length,
    timestamp: new Date().toISOString(),
    categories: {}
  };

  // 统计每个分类的数量
  allData.forEach(item => {
    if (item.Category) {
      stats.categories[item.Category] = (stats.categories[item.Category] || 0) + 1;
    }
  });

  fs.writeFileSync('./src/data/mcp-stats.json', JSON.stringify(stats, null, 2), 'utf-8');
  console.log('统计信息已保存到 ./src/data/mcp-stats.json');
};

// 执行数据获取
fetchAllPages().catch(console.error);

File: scripts/fetch_mcp_data.py
#!/usr/bin/env python3
import json
import subprocess
import time

COOKIES = 'cna=cAShIM33lxUCAXPOep8X1ti3; csrf_token=4wNSbp0Rx2qL7QIeWMYUR6r6cOU%3D; t=1a5b1f2f6a81129bca720a75c6e98490; acw_sc__v2=197d84838-b050dd5327f6d707658ffbdb2a9a0fd0583d06bf1e7c7f4d3a'
CSRF_TOKEN = '4wNSbp0Rx2qL7QIeWMYUR6r6cOU='
TOTAL_PAGES = 10  # 获取前10页数据

print("开始获取MCP服务数据...")

all_servers = []

for page in range(1, TOTAL_PAGES + 1):
    print(f"正在获取第 {page}/{TOTAL_PAGES} 页...")

    curl_cmd = [
        'curl', '-s',
        'https://modelscope.cn/api/v1/dolphin/mcpServers',
        '-X', 'PUT',
        '-H', 'Content-Type: application/json',
        '-b', COOKIES,
        '-H', f'X-CSRF-TOKEN: {CSRF_TOKEN}',
        '-H', 'User-Agent: Mozilla/5.0',
        '--data-raw', json.dumps({
            "PageSize": 30,
            "PageNumber": page,
            "Query": "",
            "Criterion": []
        })
    ]

    try:
        result = subprocess.run(curl_cmd, capture_output=True, text=True)
        response_text = result.stdout

        # 解析JSON
        data = json.loads(response_text)

        if data.get('Success') and data.get('Data', {}).get('Servers'):
            servers = data['Data']['Servers']
            all_servers.extend(servers)
            print(f"  ✓ 获取到 {len(servers)} 条数据")

            # 显示总数信息
            if page == 1:
                total_count = data['Data'].get('TotalCount', 0)
                print(f"  总共有 {total_count} 条MCP服务")
        else:
            print(f"  ✗ 第 {page} 页获取失败或无数据")
            break

    except json.JSONDecodeError as e:
        print(f"  ✗ JSON解析失败: {e}")
        break
    except Exception as e:
        print(f"  ✗ 错误: {e}")
        break

    # 延迟避免请求过快
    time.sleep(0.3)

# 保存到文件
output_path = 'src/data/mcp-all-services.json'
with open(output_path, 'w', encoding='utf-8') as f:
    json.dump(all_servers, f, ensure_ascii=False, indent=2)

print("\n" + "="*50)
print(f"✅ 数据已保存到 {output_path}")
print(f"📊 总计: {len(all_servers)} 条MCP服务")
print("="*50)

# 统计分类
categories = {}
for server in all_servers:
    cats = server.get('Category', [])
    if isinstance(cats, list):
        for cat in cats:
            categories[cat] = categories.get(cat, 0) + 1

print("\n分类统计:")
for cat, count in sorted(categories.items(), key=lambda x: x[1], reverse=True):
    print(f"  {cat}: {count}")

